// @flarum/cli/lib/steps/js/model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateModelDefinition = void 0;
const base_1 = require("./base");
class GenerateModelDefinition extends base_1.BaseJsStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate JS Model Definition';
        this.exposes = [];
    }
    async getDefinition(io) {
        const className = await io.getParam({ name: 'className', message: 'Class name', type: 'text' });
        let modelType = await io.getParam({ name: 'modelType', message: 'Model type', type: 'text' });
        modelType = modelType.includes('-') ? `['${modelType}']` : `.${modelType}`;
        return `app.store.models${modelType} = ${className};`;
    }
    async getImports(frontend, paths, io) {
        const className = await io.getParam({ name: 'className', message: 'Class name', type: 'text' });
        const classNamespace = await io.getParam({ name: 'classNamespace', message: 'Class namespace', type: 'text' });
        const importPath = this.importPath(frontend, classNamespace);
        return `import ${className} from '${importPath}';`;
    }
    getExposed(_paths, _paramProvider) {
        return {};
    }
}
exports.GenerateModelDefinition = GenerateModelDefinition;
