// @flarum/cli/lib/boilersmith/exposed-param-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExposedParamManager = void 0;
function ranKey(stepName, packagePath = '') {
    return JSON.stringify([stepName, packagePath]);
}
function paramKey(stepName, paramName, packagePath = '') {
    return JSON.stringify([stepName, paramName, packagePath]);
}
class ExposedParamManager {
    constructor() {
        this.stepsRan = new Set();
        this.exposedParams = new Map();
    }
    stepRan(stepName, packagePath) {
        return this.stepsRan.has(ranKey(stepName, packagePath));
    }
    has(stepName, paramName, packagePath) {
        return this.exposedParams.has(paramKey(stepName, paramName, packagePath));
    }
    get(stepName, paramName, packagePath) {
        return this.exposedParams.get(paramKey(stepName, paramName, packagePath));
    }
    add(stepName, params, packagePath) {
        this.stepsRan.add(ranKey(stepName, packagePath));
        Object.entries(params).forEach(([paramName, value]) => {
            this.exposedParams.set(paramKey(stepName, paramName, packagePath), value);
        });
    }
}
exports.ExposedParamManager = ExposedParamManager;
