// @flarum/cli/lib/steps/extenders/api-serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateApiSerializerAttributesExtender = void 0;
const tslib_1 = require("tslib");
const chalk_1 = (0, tslib_1.__importDefault)(require("chalk"));
const validation_1 = require("../../utils/validation");
const php_provider_1 = require("../../providers/php-provider");
const base_1 = require("./base");
class GenerateApiSerializerAttributesExtender extends base_1.BaseExtenderStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate API Serializer attributes extender';
        this.schema = {
            extenderDef: {
                extender: {
                    className: '\\Flarum\\Extend\\ApiSerializer',
                    args: [
                        {
                            type: php_provider_1.ExpressionType.CLASS_CONST,
                            value: '${serializerClass}',
                            auxiliaryValue: 'class',
                        },
                    ],
                },
                methodCalls: [
                    {
                        methodName: 'attributes',
                        args: [
                            {
                                type: php_provider_1.ExpressionType.CLOSURE,
                                value: {
                                    params: [
                                        {
                                            name: 'serializer',
                                            type: '${serializerClass}',
                                            typeType: php_provider_1.ParamTypeType.CLASS,
                                        },
                                        {
                                            name: 'model',
                                            type: '${modelClass}',
                                            typeType: php_provider_1.ParamTypeType.CLASS,
                                        },
                                        {
                                            name: 'attributes',
                                            type: 'array',
                                            typeType: php_provider_1.ParamTypeType.PRIMITIVE,
                                        },
                                    ],
                                    commentText: 'See https://docs.flarum.org/extend/api.html#serializers',
                                    return: {
                                        type: php_provider_1.ExpressionType.VARIABLE,
                                        value: 'attributes',
                                    },
                                },
                            },
                        ],
                    },
                ],
            },
            params: [
                {
                    name: 'serializerClass',
                    type: 'text',
                    validate: validation_1.Validator.class,
                    message: `API Serializer (${chalk_1.default.dim('Vendor\\Path\\Api\\Serializer\\Class')})`,
                },
                {
                    name: 'modelClass',
                    type: 'text',
                    validate: validation_1.Validator.class,
                    message: `Model (${chalk_1.default.dim('Vendor\\Path\\Model\\Class')})`,
                },
            ],
        };
    }
}
exports.GenerateApiSerializerAttributesExtender = GenerateApiSerializerAttributesExtender;
