// @flarum/cli/lib/steps/extenders/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateEventListenerExtender = void 0;
const tslib_1 = require("tslib");
const validation_1 = require("../../utils/validation");
const php_provider_1 = require("../../providers/php-provider");
const base_1 = require("./base");
const chalk_1 = (0, tslib_1.__importDefault)(require("chalk"));
class GenerateEventListenerExtender extends base_1.BaseExtenderStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate Event Listener extender';
        this.schema = {
            extenderDef: {
                extender: {
                    className: '\\Flarum\\Extend\\Event',
                },
                methodCalls: [
                    {
                        methodName: 'listen',
                        args: [
                            {
                                type: php_provider_1.ExpressionType.CLASS_CONST,
                                value: '${eventClass}',
                                auxiliaryValue: 'class',
                            },
                            {
                                type: php_provider_1.ExpressionType.CLASS_CONST,
                                value: '${listenerClass}',
                                auxiliaryValue: 'class',
                            },
                        ],
                    },
                ],
            },
            params: [
                {
                    name: 'eventClass',
                    type: 'text',
                    validate: validation_1.Validator.class,
                    message: `Event Class (${chalk_1.default.dim('Vendor\\Path\\Event')})`,
                },
                {
                    name: 'listenerClass',
                    type: 'text',
                    validate: validation_1.Validator.class,
                    message: `Your Listener Class (${chalk_1.default.dim('Your\\Extension\\Listener\\ClassName')})`,
                },
            ],
        };
    }
}
exports.GenerateEventListenerExtender = GenerateEventListenerExtender;
