// @flarum/cli/lib/steps/extenders/policy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeneratePolicyExtender = void 0;
const tslib_1 = require("tslib");
const validation_1 = require("../../utils/validation");
const php_provider_1 = require("../../providers/php-provider");
const base_1 = require("./base");
const chalk_1 = (0, tslib_1.__importDefault)(require("chalk"));
class GeneratePolicyExtender extends base_1.BaseExtenderStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate policy extender';
        this.schema = {
            extenderDef: {
                extender: {
                    className: '\\Flarum\\Extend\\Policy',
                },
                methodCalls: [
                    {
                        methodName: 'modelPolicy',
                        args: [
                            {
                                type: php_provider_1.ExpressionType.CLASS_CONST,
                                value: '${modelClass}',
                                auxiliaryValue: 'class',
                            },
                            {
                                type: php_provider_1.ExpressionType.CLASS_CONST,
                                value: '${policyClass}',
                                auxiliaryValue: 'class',
                            },
                        ],
                    },
                ],
            },
            params: [
                {
                    name: 'modelClass',
                    type: 'text',
                    validate: validation_1.Validator.class,
                    message: `Model Class (${chalk_1.default.dim('Vendor\\Path\\Model')})`,
                },
                {
                    name: 'policyClass',
                    type: 'text',
                    validate: validation_1.Validator.class,
                    message: `Your Policy Class (${chalk_1.default.dim('Your\\Extension\\Access\\ClassName')})`,
                },
            ],
        };
    }
}
exports.GeneratePolicyExtender = GeneratePolicyExtender;
