// @flarum/cli/lib/steps/extenders/service-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateServiceProviderExtender = void 0;
const tslib_1 = require("tslib");
const validation_1 = require("../../utils/validation");
const php_provider_1 = require("../../providers/php-provider");
const base_1 = require("./base");
const chalk_1 = (0, tslib_1.__importDefault)(require("chalk"));
class GenerateServiceProviderExtender extends base_1.BaseExtenderStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate Service Provider extender';
        this.schema = {
            extenderDef: {
                extender: {
                    className: '\\Flarum\\Extend\\ServiceProvider',
                },
                methodCalls: [
                    {
                        methodName: 'register',
                        args: [
                            {
                                type: php_provider_1.ExpressionType.CLASS_CONST,
                                value: '${providerClass}',
                                auxiliaryValue: 'class',
                            },
                        ],
                    },
                ],
            },
            params: [
                {
                    name: 'providerClass',
                    type: 'text',
                    validate: validation_1.Validator.class,
                    message: `Provider Class (${chalk_1.default.dim('Vendor\\Path\\Event')})`,
                },
            ],
        };
    }
}
exports.GenerateServiceProviderExtender = GenerateServiceProviderExtender;
