// @flarum/cli/lib/steps/stubs/frontend/component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateComponentStub = void 0;
const validation_1 = require("../../../utils/validation");
const js_base_1 = require("../js-base");
class GenerateComponentStub extends js_base_1.BaseJsStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate a Component';
        this.additionalExposes = ['frontend'];
        this.schema = {
            recommendedSubdir: '${frontend}/components',
            sourceFile: 'frontend/component.js',
            params: [
                {
                    name: 'frontend',
                    type: 'text',
                    message: 'Frontend name',
                },
                {
                    name: 'className',
                    type: 'text',
                    message: 'Component class name',
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
            ],
        };
    }
}
exports.GenerateComponentStub = GenerateComponentStub;
