// @flarum/cli/lib/steps/stubs/frontend/modal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateModalStub = void 0;
const validation_1 = require("../../../utils/validation");
const js_base_1 = require("../js-base");
class GenerateModalStub extends js_base_1.BaseJsStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate Modal Component';
        this.additionalExposes = ['frontend'];
        this.schema = {
            recommendedSubdir: '${frontend}/components',
            sourceFile: 'frontend/modal.js',
            params: [
                {
                    name: 'frontend',
                    type: 'text',
                    message: 'Frontend name',
                },
                {
                    name: 'className',
                    type: 'text',
                    message: 'Modal class name',
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
            ],
        };
    }
}
exports.GenerateModalStub = GenerateModalStub;
