// @flarum/cli/lib/steps/stubs/frontend/model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateModelStub = void 0;
const validation_1 = require("../../../utils/validation");
const js_base_1 = require("../js-base");
const model_name_1 = require("../../../utils/model-name");
class GenerateModelStub extends js_base_1.BaseJsStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate Model Class';
        this.additionalExposes = ['frontend', 'modelType', 'classNamespace'];
        this.schema = {
            recommendedSubdir: '${frontend}/models',
            sourceFile: 'frontend/model.js',
            params: [
                {
                    name: 'frontend',
                    type: 'text',
                    message: 'Frontend name',
                },
                {
                    name: 'className',
                    type: 'text',
                    message: 'Model class name',
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
            ],
        };
    }
    async compileParams(fs, paths, io) {
        const params = await super.compileParams(fs, paths, io);
        params.modelType = (0, model_name_1.pluralKebabCaseModel)(params.className);
        return params;
    }
}
exports.GenerateModelStub = GenerateModelStub;
