// @flarum/cli/lib/commands/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const composer_1 = require("../steps/misc/composer");
const yarn_1 = require("../steps/misc/yarn");
const base_command_1 = (0, tslib_1.__importDefault)(require("../base-command"));
const yosay_1 = (0, tslib_1.__importDefault)(require("yosay"));
const gen_ext_scaffolder_1 = require("../steps/gen-ext-scaffolder");
const git_1 = require("../steps/misc/git");
const npm_1 = require("../steps/misc/npm");
class Init extends base_command_1.default {
    constructor() {
        super(...arguments);
        this.requireExistingExtension = false;
    }
    steps(stepManager, extRoot) {
        const packageManager = this.jsPackageManager(extRoot);
        return stepManager
            .namedStep('gitInit', new git_1.GitInit(), { optional: true, confirmationMessage: 'Run `git init`? (recommended)', default: true })
            .namedStep('skeleton', (0, gen_ext_scaffolder_1.genExtScaffolder)().genInitStep())
            .step(new composer_1.ComposerInstall(), { optional: true, confirmationMessage: 'Run `composer install`? (recommended)', default: true })
            .step(packageManager === 'npm' ? new npm_1.NpmInstall() : new yarn_1.YarnInstall(), { optional: true, confirmationMessage: `Run \`${packageManager !== null && packageManager !== void 0 ? packageManager : 'yarn'}\`? (recommended)`, default: true }, [
            {
                sourceStep: 'skeleton',
                exposedName: 'modules.js',
                dontRunIfFalsy: true,
            },
        ]);
    }
    welcomeMessage() {
        return (0, yosay_1.default)('Welcome to a Flarum extension generator\n\n- Flarum Team');
    }
    goodbyeMessage() {
        return 'Extension generation complete! Visit https://docs.flarum.org/extend to learn more about Flarum extension development.';
    }
    async additionalPreRunChecks(extRoot) {
        if (await this.confirmOverrideFiles(extRoot, '**/*', 'Directory not empty. Overwrite?')) {
            this.deleteFiles(extRoot, '**/*');
        }
    }
}
exports.default = Init;
Init.description = 'Create a new Flarum extension';
Init.flags = Object.assign({}, base_command_1.default.flags);
Init.args = [...base_command_1.default.args];
