// @flarum/cli/lib/steps/stubs/backend/api-serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateApiSerializerStub = void 0;
const tslib_1 = require("tslib");
const chalk_1 = (0, tslib_1.__importDefault)(require("chalk"));
const validation_1 = require("../../../utils/validation");
const model_name_1 = require("../../../utils/model-name");
const php_base_1 = require("../php-base");
class GenerateApiSerializerStub extends php_base_1.BasePhpStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate Api Serializer Class';
        this.additionalExposes = ['className', 'modelClass', 'modelClassName'];
        this.additionalImplicitParams = ['modelType'];
        this.phpClassParams = ['modelClass'];
        this.schema = {
            recommendedSubdir: 'Api/Serializer',
            sourceFile: 'backend/api-serializer.php',
            params: [
                {
                    name: 'className',
                    type: 'text',
                    message: 'Serializer class name',
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
                {
                    name: 'modelClass',
                    type: 'text',
                    message: `Model Class (${chalk_1.default.dim('Vendor\\Path\\Model')})`,
                    validate: validation_1.Validator.class,
                },
                {
                    name: 'modelClassName',
                    type: 'text',
                    message: 'Model class name',
                },
                {
                    name: 'modelType',
                    type: 'text',
                    message: 'Model type',
                },
            ],
        };
    }
    async compileParams(fs, paths, io) {
        const params = await super.compileParams(fs, paths, io);
        params.modelType = (0, model_name_1.pluralKebabCaseModel)(params.modelClassName);
        return params;
    }
}
exports.GenerateApiSerializerStub = GenerateApiSerializerStub;
