// @flarum/cli/lib/steps/stubs/backend/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateCommandStub = void 0;
const tslib_1 = require("tslib");
const chalk_1 = (0, tslib_1.__importDefault)(require("chalk"));
const validation_1 = require("../../../utils/validation");
const php_base_1 = require("../php-base");
class GenerateCommandStub extends php_base_1.BasePhpStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate a command class';
        this.additionalExposes = [];
        this.phpClassParams = [];
        this.schema = {
            recommendedSubdir: 'Console',
            sourceFile: 'backend/command.php',
            params: [
                {
                    name: 'className',
                    type: 'text',
                    message: 'Command class name',
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
                {
                    name: 'commandName',
                    type: 'text',
                    message: `Command name (${chalk_1.default.dim('namespace:command')})`,
                    validate: validation_1.Validator.commandName,
                },
                {
                    name: 'commandDescription',
                    type: 'text',
                    message: 'Command description',
                },
            ],
        };
    }
}
exports.GenerateCommandStub = GenerateCommandStub;
