// @flarum/cli/lib/steps/stubs/backend/event-listener.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateEventListenerStub = void 0;
const tslib_1 = require("tslib");
const chalk_1 = (0, tslib_1.__importDefault)(require("chalk"));
const validation_1 = require("../../../utils/validation");
const php_base_1 = require("../php-base");
class GenerateEventListenerStub extends php_base_1.BasePhpStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate Event Listener Class';
        this.additionalExposes = ['eventClass'];
        this.phpClassParams = ['eventClass'];
        this.schema = {
            recommendedSubdir: 'Listener',
            sourceFile: 'backend/event-listener.php',
            params: [
                {
                    name: 'className',
                    type: 'text',
                    message: 'Listener class name',
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
                {
                    name: 'eventClass',
                    type: 'text',
                    message: `Event Class (${chalk_1.default.dim('Vendor\\Path\\Event')})`,
                    validate: validation_1.Validator.class,
                },
                {
                    name: 'eventClassName',
                    type: 'text',
                    message: 'Event Class Name',
                },
            ],
        };
    }
}
exports.GenerateEventListenerStub = GenerateEventListenerStub;
