// @flarum/cli/lib/steps/stubs/backend/handler-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateHandlerCommandStub = void 0;
const tslib_1 = require("tslib");
const chalk_1 = (0, tslib_1.__importDefault)(require("chalk"));
const validation_1 = require("../../../utils/validation");
const php_base_1 = require("../php-base");
class GenerateHandlerCommandStub extends php_base_1.BasePhpStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate Handler Class';
        this.additionalExposes = ['className', 'classType'];
        this.phpClassParams = [];
        this.schema = {
            recommendedSubdir: 'Command',
            sourceFile: 'backend/handler/command.php',
            params: [
                {
                    name: 'className',
                    type: 'text',
                    message: `Handler Command class name (example: ${chalk_1.default.dim('EditModel')})`,
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
                {
                    name: 'classType',
                    type: 'autocomplete',
                    message: 'Class Type',
                    choices: ['None', 'Create', 'Update', 'Delete'].map((type) => ({
                        title: type,
                        value: type.toLowerCase(),
                    })),
                },
            ],
        };
    }
}
exports.GenerateHandlerCommandStub = GenerateHandlerCommandStub;
