// @flarum/cli/lib/steps/stubs/backend/integration-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateIntegrationTestStub = void 0;
const validation_1 = require("../../../utils/validation");
const php_base_1 = require("../php-base");
class GenerateIntegrationTestStub extends php_base_1.BasePhpStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate Integration Test Class';
        this.additionalExposes = [];
        this.phpClassParams = [];
        this.schema = {
            root: './tests',
            recommendedSubdir: 'integration.api',
            sourceFile: 'backend/integration-test.php',
            params: [
                {
                    name: 'className',
                    type: 'text',
                    message: 'Test class name',
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
                {
                    name: 'extensionId',
                    type: 'text',
                    message: 'Extension Id',
                },
            ],
        };
    }
}
exports.GenerateIntegrationTestStub = GenerateIntegrationTestStub;
