// @flarum/cli/lib/steps/stubs/backend/policy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeneratePolicyStub = void 0;
const tslib_1 = require("tslib");
const chalk_1 = (0, tslib_1.__importDefault)(require("chalk"));
const validation_1 = require("../../../utils/validation");
const php_base_1 = require("../php-base");
class GeneratePolicyStub extends php_base_1.BasePhpStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate a policy Class';
        this.additionalExposes = ['modelClass', 'modelClassName'];
        this.phpClassParams = ['modelClass'];
        this.schema = {
            recommendedSubdir: 'Access',
            sourceFile: 'backend/policy.php',
            params: [
                {
                    name: 'className',
                    type: 'text',
                    message: 'Policy class name',
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
                {
                    name: 'modelClass',
                    type: 'text',
                    message: `Model Class (${chalk_1.default.dim('Vendor\\Path\\Model')})`,
                    validate: validation_1.Validator.class,
                },
                {
                    name: 'modelClassName',
                    type: 'text',
                    message: 'Class Namespace',
                },
            ],
        };
    }
}
exports.GeneratePolicyStub = GeneratePolicyStub;
