// @flarum/cli/lib/steps/stubs/backend/service-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateServiceProviderStub = void 0;
const validation_1 = require("../../../utils/validation");
const php_base_1 = require("../php-base");
class GenerateServiceProviderStub extends php_base_1.BasePhpStubStep {
    constructor() {
        super(...arguments);
        this.type = 'Generate Service Provider Class';
        this.schema = {
            recommendedSubdir: '',
            sourceFile: 'backend/service-provider.php',
            params: [
                {
                    name: 'className',
                    type: 'text',
                    message: 'Provider class name',
                    validate: validation_1.Validator.className,
                },
                {
                    name: 'classNamespace',
                    type: 'text',
                    message: 'Class Namespace',
                },
            ],
        };
    }
}
exports.GenerateServiceProviderStub = GenerateServiceProviderStub;
